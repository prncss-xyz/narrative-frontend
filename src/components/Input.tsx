import { Box, InputRaw } from "./basics";

// type="number" behehaves too weird
export function Input({
  placeholder,
  value,
  setValue,
  disabled,
  ...props
}: {
  placeholder: string;
  value: string;
  setValue: (value: string) => void;
  disabled?: boolean;
  [prop: string]: unknown; // TODO: could be more restrictive
}) {
  const handleChange = ({
    currentTarget: { value },
  }: React.ChangeEvent<HTMLInputElement>) => {
    setValue(value);
  };
  return (
    <>
      {disabled ? (
        <Box display="inline" {...props}>
          {String(value)}
        </Box>
      ) : (
        <InputRaw
          type="text"
          placeholder={placeholder}
          value={value}
          onChange={handleChange}
          {...props}
        />
      )}
    </>
  );
}
