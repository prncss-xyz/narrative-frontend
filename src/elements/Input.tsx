export function identityString(valueString: string) {
  return valueString;
}

export function positve(valueString: string) {
  const value = Number(valueString);
  if (value >= 0) return value;
  return;
}

// type="number" behehaves too weird
export function Input<T>({
  placeholder,
  value,
  setValue,
  disabled,
  convert,
}: {
  placeholder: string;
  value: T;
  setValue: (value: T) => void;
  disabled?: boolean;
  convert: (value: string) => T | undefined;
}) {
  const handleChange = ({
    currentTarget: { value },
  }: React.ChangeEvent<HTMLInputElement>) => {
    const value_ = convert(value);
    if (value_) setValue(value_);
  };
  return (
    <>
      {disabled ? (
        <div css={{ display: "inline" }}>{String(value)}</div>
      ) : (
        <input
          type="text"
          placeholder={placeholder}
          value={String(value)}
          onChange={handleChange}
        />
      )}
    </>
  );
}
