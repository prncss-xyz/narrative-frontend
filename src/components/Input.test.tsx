import { fireEvent, render, screen } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";
import { Input } from "./Input";

describe("Input", () => {
  describe("validate: true", () => {
    const validate = (_: string) => true;
    const setValue = vi.fn();
    const { container } = render(
      <Input
        placeholder="placeholder"
        value="value: true"
        setValue={setValue}
        validate={validate}
      />
    );
    const input = screen.getByDisplayValue("value: true");
    fireEvent.change(input, { target: { value: "new" } });
    it("should trigger when clicked", () => {
      expect(setValue).toHaveBeenCalledTimes(1);
    });
    it("should match snapshot", () => {
      expect(container).toMatchSnapshot();
    });
  });
  describe("validate: false", () => {
    const validate = (_: string) => false;
    const setValue = vi.fn();
    render(
      <Input
        placeholder="placeholder"
        value="value: false"
        setValue={setValue}
        validate={validate}
      />
    );
    const input = screen.getByDisplayValue("value: false");
    fireEvent.change(input, { target: { value: "new" } });
    it("should trigger when clicked", () => {
      expect(setValue).toHaveBeenCalledTimes(0);
    });
  });
  describe("disabled", () => {
    const validate = (_: string) => false;
    const setValue = (_: string) => {
      /* do nothing */
    };
    const { container } = render(
      <Input
        placeholder="placeholder"
        value="disabled"
        setValue={setValue}
        validate={validate}
      />
    );
    it("should match snapshot", () => {
      expect(container).toMatchSnapshot();
    });
  });
});
