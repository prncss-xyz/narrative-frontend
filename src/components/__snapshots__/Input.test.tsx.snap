// Vitest Snapshot v1

exports[`Input > disabled > should match snapshot 1`] = `<div />`;

exports[`Input > validate: true > should match snapshot 1`] = `<div />`;
